packages/react-instantsearch/src/components/Highlighter.js

```diff
@@ -1,35 +1,75 @@
 import PropTypes from 'prop-types';
 import React from 'react';
+import classNames from './classNames.js';
+
+const cx = classNames('Highlight');
+
+export const Highlight = ({
+  value,
+  highlightedTagName,
+  isHighlighted,
+  nonHighlightedTagName,
+}) => {
+  const TagName = isHighlighted ? highlightedTagName : nonHighlightedTagName;
+  const className = isHighlighted ? 'highlighted' : 'nonHighlighted';
+  return <TagName {...cx(className)}>{value}</TagName>;
+};
+
+Highlight.propTypes = {
+  value: PropTypes.string.isRequired,
+  isHighlighted: PropTypes.bool.isRequired,
+  highlightedTagName: PropTypes.string.isRequired,
+  nonHighlightedTagName: PropTypes.string.isRequired,
+};
 
 export default function Highlighter({
   hit,
   attributeName,
   highlight,
   highlightProperty,
   tagName,
+  nonHighlightedTagName,
+  separator,
 }) {
   const parsedHighlightedValue = highlight({
     hit,
     attributeName,
     highlightProperty,
   });
-  const reactHighlighted = parsedHighlightedValue.map((v, i) => {
-    const key = `split-${i}-${v.value}`;
-    if (!v.isHighlighted) {
-      return (
-        <span key={key} className="ais-Highlight__nonHighlighted">
-          {v.value}
-        </span>
-      );
-    }
-    const HighlightedTag = tagName ? tagName : 'em';
-    return (
-      <HighlightedTag key={key} className="ais-Highlight__highlighted">
-        {v.value}
-      </HighlightedTag>
-    );
-  });
-  return <span className="ais-Highlight">{reactHighlighted}</span>;
+
+  return (
+    <span className="ais-Highlight">
+      {parsedHighlightedValue.map((item, i) => {
+        if (Array.isArray(item)) {
+          const isLast = i === parsedHighlightedValue.length - 1;
+          return (
+            <span key={`split-${i}-${hit[attributeName][i]}`}>
+              {item.map((element, index) => (
+                <Highlight
+                  key={`split-${index}-${element.value}`}
```

Comment:
The key generation is slightly different, since in one case we're accessing an element of an array (`hit[attributeName][i]`) and for the case where `attributeName` is a string, we do `element.value`. Do I do a function a maybe pass an attribute which will be a Boolean to specify whether it is an array and return either the first option or the second or leave it as is?

Reply:
Nope, you can simply create function that takes `(index, value)` as parameters. Then you called it with the according values (ex: `element.value`, `hit[attributeName][i]`).